operationalize creating initial config file

gouthamr321 opened the pull request:
This just takes chris's original script and operationalizes it and adds a few changes

notable changes:
- use an argument parser so things arent hard coded
- includes a helper function to sort round names
- Saves config file in json format instead of as a yaml format. This is because yaml files don't preserve the order of keys in a dictionary. See here: [https://github.com/yaml/pyyaml/issues/110](here). I'm not sure if saving this will cause issues downstream..... though should be easy to modify other scripts to read in the json file instead)

The output config files are here: /allen/aics/assay-dev/users/Goutham/4iProcessing-/4iprocessing/json_configs....... These look consistent to me compared to the original yaml files, but I'm hoping @cfrick13 can confirm

gouthamr321 committed:
operationalize creating initial config file

cfrick13 commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,130 @@
+    # function to sort rounds
+    Orginal_numbered_list = []
+    for i in range(len(rounds)):
+        int(re.search(r'\d+', rounds[i]).group())
```
this line is superfluous i think

gouthamr321 replied:
Nice catch, line 24 is superfluous

cfrick13 commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,130 @@
+                    # the reference image type (e.g. nuclear dye or brightfield) should be the consisently chosen for all rounds
+                    # if the image is a timelapse or round01 then choose the reference channel to be the last channel in the image set , =-1.
+
+                    ref_channel = -2
```
lines 74 - 79 should be moved below line 84

gouthamr321 replied:
This doesn't make any functional changes.... we are just defining the reference channel which doesn't affect the reading of the file..... but maybe I'm not understanding?

cfrick13 replied:
yes. no functional changes. it just makes sense to put the code where we choose channels after we have determined the channels from reader.channels. so its just a cosmetic change.

cfrick13 commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,130 @@
+                    # if the image is a timelapse or round01 then choose the reference channel to be the last channel in the image set , =-1.
+
+                    ref_channel = -2
+                    if bool(re.search('time|1(?![0-9])',round_num, re.IGNORECASE)) and round_num!="Round 11": 
```
need to make this catch any round greater than 10. right now it only looks for round 11 but in theory we may have 12 or 13 rounds in the future.

gouthamr321 replied:
Just confirming, here we only what this if statement to be executed if we are looking at the first round or the 20X timelapse.... is that correct?

cfrick13 replied:
correct

cfrick13 commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,130 @@
+                    detailid={}
+                    detailid['round'] = round_num
+                    detailid['item'] = 'czi'
+                    detailid['zchannels'] = zchannels
```
change name to `channels` instead of `zchannels`

cfrick13 replied:
I also want channels to be at the bottom, so maybe move this line below 107.

gouthamr321 replied:
ohh, nice catch

cfrick13 commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,130 @@
+            config['barcode'] = barcode
+            config['scope'] = scope
+            
+            # output path defies folder where all images get stored after they get processed
```
typo: "defies"--> "defines"

gouthamr321 replied:
Addressed in new commits

cfrick13 reviewed (commented):
made some comments in the code. looks good overall

gouthamr321 committed:
incorporate feedback from chris

gouthamr321 committed:
add some documentation for the functions used

gouthamr321 committed:
renamed file

cfrick13 commented:
we talked and decided to wait to merge this code until we finish cleaning up variable names in the code (in `FormattingAndSorting` branch). And also decided to look at other ways of writing yamls to violate alphabetical order

aditnath commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,144 @@
+# this code only helps make the yaml files...it does not generate a perfect yaml automatically. 
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--input_dirs', type=str, default=["/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000733","/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000724", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000728", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000726", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000725"], help="input dirs to ...
```
might want to not have these hardcoded as these will change once the isilon is gone?

gouthamr321 replied:
nice catch, took care of this

aditnath commented on 4iprocessing/generate_initial_config.py:
```diff
@@ -0,0 +1,144 @@
+            print("round list is {}".format(round_list))
+            for round_num in round_list:
+                ppath = os.path.join(pdir, round_num)
+                czi_list = [x for x in os.listdir(ppath) if ('.czi' in x)&bool(re.search('20x',x,re.IGNORECASE))]
```
we will need to include support for ome.tiffs as well, since those are the aligned files

gouthamr321 replied:
also just took care of this one

aditnath reviewed (changes requested):
left a couple comments, we might want to address the ome tiff functionality before merging

gouthamr321 committed:
maintain order of dict elements in yaml file

gouthamr321 committed:
support for tiff and remove arg parse defaults

gouthamr321 committed:
fix for parser issue not reading in command line

gouthamr321 committed:
delete debug print statements

gouthamr321 commented:
Incorporated some changes since last time. Notable changes are:
- workaround to save config file in yaml format without messing up order of elements in dictionary
- defined some operations as separate functions(e.g. determining which scenes to toss, determining reference channel)
- support for files saved in .tiff format for the future
- change to argument parser to not hardcode input directory files 

Not sure if this is ready to merge or if we still need to wait for FormattingAndSorting branch to be merged first? Thoughts?

cfrick13 commented:
We should be ready to merge this now. FormattingAndSorting branch is merged

gouthamr321 committed:
merge
Merge branch 'main' of https://github.com/aics-int/4iProcessing- into goutham_dev

cfrick13 reviewed (approved):
looks good. Might just need change requests marked as completed

gouthamr321 commented:
Operationalize config file generation
